File: src/main/java/com/thoughtworks/ddd/ParkingBoy.java
package com.thoughtworks.ddd;

import java.util.List;
import java.util.stream.Collectors;

public class ParkingBoy {

    private List<ParkingLot> parkingLots;
    private SearchParkingLotRule rule;

    public ParkingBoy(List<ParkingLot> parkingLots, SearchParkingLotRule rule) {
        this.parkingLots = parkingLots;
        this.rule = rule;
    }

    public Receipt parkCar(Car car) {
        ParkingLot parkingLot = rule.search(parkingLots);
        try {
            return parkingLot.parkCar(car);
        } catch (ParkException e) {
            e.printStackTrace();
            return null;
        }
    }

    public ParkingLot searchAvailableParkingLot() {
        List<ParkingLot> availableParkingLots = parkingLots.stream().filter(ParkingLot::isNotFull).collect(Collectors.toList());
        if (availableParkingLots.isEmpty()) {
            return null;
        }
        return availableParkingLots.get(0);
    }

    public Boolean isAvailable() {
        return parkingLots.stream().anyMatch(ParkingLot::isNotFull);
    }
}

File: src/main/java/com/thoughtworks/ddd/domain/ParkingBoy.kt
package com.thoughtworks.ddd.domain

import com.thoughtworks.ddd.infrastructure.ParkException

class ParkingBoy(
        private val parkingLots: List<ParkingLot>,
        private val parkingPolicy: ParkingPolicy
) {

    val isAvailable = parkingLots.any { it.isNotFull }

    fun findAvailableParkingLot(): ParkingLot =
            parkingPolicy.findParkingLot(parkingLots)
                    ?: throw ParkException("当前无可用停车场")

}